### PR TITLE
Kernel transform (for #185)

### PR DESCRIPTION
**Pull request by arporter**
Implements a `Kern.ast` property that creates and returns the fparser2 AST of the kernel.
Implements a new, `ACCRoutineTrans` transformation that uses the above functionality to modify a kernel by inserting a `!$acc routine` directive.

This PR does *not* implement support for writing the transformed kernel back out to file. That is the subject of #201.

### CODE REVIEW

File: doc/transformations.rst [hunk 1]
```diff
@@ -119,6 +119,16 @@ can be found in the API-specific sections).
 	  :ref:`MoveTrans <sec_move_trans>` transformation may be used
           for this.
 
+Kernels
+-------
+
+Occasionally it is not sufficient to transform only the generated
```
Comment (rupertford) [inline doc/transformations.rst]: I don't necessarily think that it is going to only be occasionally in the future so I would be a bit more bullish here and say something like ...
"PSyclone supports the transformation of Kernels as well as PSy-layer code. Currently ..."

File: src/psyclone/f2pygen.py [hunk 2]
```diff
@@ -583,7 +583,7 @@ def __init__(self, parent, content):
         :type parent: :py:class:`psyclone.f2pygen.BaseGen`
         :param str content: the content of the comment
         '''
-        reader = FortranStringReader("! content\n")
+        reader = FortranStringReader("! content\n", ignore_comments=False)
```
Comment (rupertford) [inline src/psyclone/f2pygen.py]: Is this change required due to you adding support for comments recently and defaulting ignore to True? If so, how did we get away with it until now?

File: src/psyclone/f2pygen.py [hunk 3]
```diff
@@ -612,7 +612,7 @@ def __init__(self, parent, language, position, directive_type, content):
         self._language = language
         self._directive_type = directive_type
 
-        reader = FortranStringReader("! content\n")
+        reader = FortranStringReader("! content\n", ignore_comments=False)
```
Comment (rupertford) [inline src/psyclone/f2pygen.py]: See previous comment.

File: src/psyclone/psyGen.py [hunk 4]
```diff
@@ -2803,12 +2803,27 @@ def gen_code(self, parent):
 
 
 class Kern(Call):
+    '''
+    Class representing a Kernel call within the Schedule (AST) of an Invoke.
```
Comment (rupertford) [inline src/psyclone/psyGen.py]: I'd not thought of a Schedule being as AST before but I guess it is.

File: src/psyclone/psyGen.py [hunk 5]
```diff
@@ -2803,12 +2803,27 @@ def gen_code(self, parent):
...
+
+    :param type KernelArguments: the sub-class of KernelArguments to create.
```
Comment (rupertford) [inline src/psyclone/psyGen.py]: Is this not a more specific type of class?

File: src/psyclone/psyGen.py [hunk 6]
```diff
@@ -2934,6 +2949,26 @@ def is_coloured(self):
...
+        from fparser.two import Fortran2003
+        # If we've already got the AST then just return it
+        if self._fp2_ast:
+            return self._fp2_ast
+        # Use the fparser1 AST to generate Fortran source
+        fortran = self._module_code.tofortran()
+        # Parse that Fortran using fparser2
+        reader = FortranStringReader(fortran)
+        self._fp2_ast = Fortran2003.Program(reader)
+        return self._fp2_ast
```
Comment (rupertford) [inline src/psyclone/psyGen.py]: Nicely done.

File: src/psyclone/transformations.py [hunk 7]
```diff
@@ -2262,3 +2262,82 @@ def apply(self, sched):
         schedule.addchild(data_dir, index=0)
 
         return schedule, keep
+
+
+class ACCRoutineTrans(Transformation):
+    '''
+    Transform a kernel subroutine through the addition of a "!$acc routine"
```
Comment (rupertford) [inline src/psyclone/transformations.py]: Change "through the addition of" to "by adding"

File: src/psyclone/transformations.py [hunk 8]
```diff
@@ -2262,3 +2262,82 @@ def apply(self, sched):
...
+    directive (causing it to be compiled for the OpenACC accelerator device).
+    For example:
+
+    >>> from psyclone.parse import parse
+    >>> from psyclone.psyGen import PSyFactory
+    >>> api = "gocean1.0"
+    >>> filename = "nemolite2d_alg.f90"
+    >>> ast, invokeInfo = parse(filename, api=api, invoke_name="invoke")
```
Comment (rupertford) [inline src/psyclone/transformations.py]: Can we drop the invoke_name argument here as it defaults to "invoke"?

File: src/psyclone/transformations.py [hunk 9]
```diff
@@ -2262,3 +2262,82 @@ def apply(self, sched):
...
+    >>> psy = PSyFactory(api).create(invokeInfo)
+    >>>
+    >>> from psyclone.transformations import ACCRoutineTrans
```
Comment (rupertford) [inline src/psyclone/transformations.py]: Just a thought ... it might be better to separate transformations on kernel, psy and algorithm layers in the future ... e.g. this could then become "from psyclone.transformations.kernel import ACCRoutineTrans"

File: src/psyclone/transformations.py [hunk 10]
```diff
@@ -2262,3 +2262,82 @@ def apply(self, sched):
...
+        for idx, node in enumerate(spec.content):
+            if not (isinstance(node, (Implicit_Part, Type_Declaration_Stmt))):
+                break
+        # Create the directive and insert it
+        cmt = Comment(FortranStringReader("!$acc routine",
+                                          ignore_comments=False))
+        spec.content.insert(idx, cmt)
+
+        # Return the now modified kernel
+        return kern, keep
```
Comment (rupertford) [inline src/psyclone/transformations.py]: Nice.

File: src/psyclone/transformations.py [hunk 11]
```diff
@@ -2262,3 +2262,82 @@ def apply(self, sched):
...
+    def name(self):
+        '''
+        :returns: the name of this transformation class.
+        :rtype: str
+        '''
+        return "ACCRoutineTrans"
+
+    def apply(self, kern):
+        '''
+        :param kern: The kernel object to transform.
```
Comment (rupertford) [inline src/psyclone/transformations.py]: I think the 'apply' method should have a (concise) description of what it does.

File: src/psyclone/tests/psyGen_test.py [hunk 12]
```diff
@@ -2264,3 +2264,18 @@ def test_find_w_args_multiple_deps():
...
+    from fparser.two import Fortran2003
+    _, invoke = get_invoke(os.path.join("gocean1p0",
+                                        "nemolite2d_alg_mod.f90"),
+                           api="gocean1.0", idx=0)
+    sched = invoke.schedule
+    kern = sched.children[0].children[0].children[0]
+    assert isinstance(kern, GOKern)
+    assert kern.ast
+    assert isinstance(kern.ast, Fortran2003.Program)
+
```
Comment (rupertford) [inline src/psyclone/tests/psyGen_test.py]: Perhaps add a check that the subroutine name is the one we were expecting?

File: src/psyclone/tests/kernel_transformation_tests.py [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+    mod = ast.content[0]
+    # Find the subroutine statement
+    for child in mod.content:
+        if isinstance(child, fparser.one.block_statements.Subroutine):
+            sub = child
+    # Find the end subroutine statement
+    for child in sub.content:
+        if isinstance(child, fparser.one.block_statements.EndSubroutine):
+            end = child
+    sub.name = "some_other_name"
```
Comment (rupertford) [inline src/psyclone/tests/kernel_transformation_tests.py]: this might leave a modified kernel ast which could affect future tests. Should we not put it back into its original state, or use monkeypatch

File: src/psyclone/tests/kernel_transformation_tests.py [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+    new_kern, _ = rtrans.apply(kern)
+    # The transformation should have populated the fparser2 AST of
+    # the kernel...
+    assert new_kern._fp2_ast
+    assert isinstance(new_kern._fp2_ast, Fortran2003.Program)
+    # Check AST contains directive
+    comments = Fortran2003.walk_ast(new_kern._fp2_ast.content,
+                                    [Fortran2003.Comment])
+    assert len(comments) == 1
+    assert str(comments[0]) == "!$acc routine"
```
Comment (rupertford) [inline src/psyclone/tests/kernel_transformation_tests.py]: Test that it has inserted it in the expected (correct) place.

**Comment by rupertford** [comment]
Please also address the conflict and bring the branch up-to-date before 2nd review

File: doc/developers.rst [hunk 15]
```diff
@@ -794,6 +794,28 @@ encapsulation for API-specific options. They do not sub-class ``Config``
 directly but store a reference back to the ``Config`` object to which they
 belong.
 
+transformations
+===============
+
+It should come as no surprise that the transformations module holds the
```
[the remark this answers is not in the document]
Comment (arporter) [reply doc/developers.rst]: Done.

File: doc/transformations.rst (hunk 1 above)
Comment (arporter) [reply doc/transformations.rst]: Done.

File: src/psyclone/f2pygen.py (hunk 2 above)
Comment (arporter) [reply src/psyclone/f2pygen.py]: I think fparser1 already had support for comments and did so by default. It's only with my recent changes for fparser2 that this has changed - these changes are on master but not in the 0.0.7 release of fparser. This ticket needs the master branch of fparser2 because it needs `Comment` support when adding an OpenACC directive to kernel source.

File: src/psyclone/psyGen.py (hunk 5 above)
Comment (arporter) [reply src/psyclone/psyGen.py]: No, it is a `type`, not an object. However, I've extended the description to make it clear that the type will be a sub-class of `psyGen.Arguments`.

File: src/psyclone/transformations.py (hunk 7 above)
Comment (arporter) [reply src/psyclone/transformations.py]: Done.

File: src/psyclone/transformations.py (hunk 8 above)
Comment (arporter) [reply src/psyclone/transformations.py]: Oh yes; done.

File: src/psyclone/transformations.py (hunk 9 above)
Comment (arporter) [reply src/psyclone/transformations.py]: Good point. I've created #211 for this.

File: src/psyclone/transformations.py (hunk 11 above)
Comment (arporter) [reply src/psyclone/transformations.py]: Done.

File: src/psyclone/tests/psyGen_test.py (hunk 12 above)
Comment (arporter) [reply src/psyclone/tests/psyGen_test.py]: Actually, `kern.ast` holds the AST for the whole Fortran module so I don't think that that check is relevant.

File: src/psyclone/tests/kernel_transformation_tests.py (hunk 13 above)
Comment (arporter) [reply src/psyclone/tests/kernel_transformation_tests.py]: Good point. I've changed it to use monkeypatch.

File: src/psyclone/tests/kernel_transformation_tests.py (hunk 14 above)
Comment (arporter) [reply src/psyclone/tests/kernel_transformation_tests.py]: Good point. Done.

File: src/psyclone/f2pygen.py (hunk 2 above)
Comment (arporter) [reply src/psyclone/f2pygen.py]: Just to complete what I said above, this means that we will need a new release of fparser before this PR can be merged to master. This will have implications for @MatthewHambley and/or @TeranIvy as the API for fparser2 has changed (I've just been burnt by that in addressing this review).

**Comment by arporter** [comment]
I think I've addressed all of the points from the first review. Ready for a second round...

File: src/psyclone/f2pygen.py (hunk 2 above)
Comment (rupertford) [reply src/psyclone/f2pygen.py]: OK, thanks for the info. I think we should change the default to include comments but that is a different issue. Regarding requiring a new release of fparser, we have discussed this and have a new strategy for supporting concurrent development of psyclone and fparser, see pr #213

File: src/psyclone/psyGen.py (hunk 5 above)
Comment (rupertford) [reply src/psyclone/psyGen.py]: You would have thought I would know that, seeing as I wrote the code - doh! Anyway the improved description is helpful.

File: src/psyclone/transformations.py (hunk 9 above)
Comment (rupertford) [reply src/psyclone/transformations.py]: Thanks

File: src/psyclone/tests/psyGen_test.py (hunk 12 above)
Comment (rupertford) [reply src/psyclone/tests/psyGen_test.py]: OK.

**Comment by rupertford** [comment]
Back to @arporter.

**Comment by arporter** [comment]
Have brought branch up-to-date with master. Tests on Travis now pass.
Re-named and corrected the `kernel_transformation_tests.py` file.
Hopefully that should be it. Ready for next review...
